### PR TITLE
build: download wintools executables separately

### PR DESCRIPTION
**Pull request by dundargoc**
The wintools executables are stored in a zip file, making it
inconvenient to bump these during releases. Instead, unpack the
executables in the deps repository and download each executable
separately.

### CODE REVIEW

**Comment by dundargoc** [comment]
@justinmk very rough draft, something like this should work. `deps` will have the following structure: https://github.com/dundargoc/deps/tree/master/opt

Will need to clean this up, this is the best I could come up with due to potato laptop limitations.

**Comment by dundargoc** [comment]
Actually unsure if this is even needed. We should be able to download the entire repo as an archive from github directly.

File: cmake.deps/CMakeLists.txt [hunk 1]
```diff
@@ -162,9 +162,29 @@ set(LUV_SHA256 16720d17b9e6d42a7008b902207ea2d61e3eeaef3cf358398dbbf3777867ec09)
 set(LUA_COMPAT53_URL https://github.com/keplerproject/lua-compat-5.3/archive/v0.9.tar.gz)
 set(LUA_COMPAT53_SHA256 ad05540d2d96a48725bb79a1def35cf6652a4e2ec26376e2617c8ce2baa6f416)
 
-# cat.exe curl.exe curl-ca-bundle.crt diff.exe tee.exe xxd.exe
-set(WINTOOLS_URL https://github.com/neovim/deps/raw/c1e7dd8de9e1b18d11dcfa0a192cd029262e5303/opt/win32tools.zip)
-set(WINTOOLS_SHA256 3c4c490a3d392ceeb1347cb77cc821a31900b688a2189276d3a1131a3f21daf1)
+set(CAT_URL https://github.com/dundargoc/deps/raw/master/opt/cat.exe)
+set(CAT_SHA256 93b8d307bb15af3968920bdea3beb869a49d166f9164853c58a4e6ffdcae61c6)
+
```
Comment (justinmk) [inline cmake.deps/CMakeLists.txt:167]: nit: let's group these all together (no blank lines) and add a comment like `// Windows only: curl.exe, xxd.exe, etc.`

File: cmake.deps/CMakeLists.txt [hunk 2]
```diff
@@ -263,8 +283,93 @@ endif()
...
+  ExternalProject_Add(cat
+    URL ${CAT_URL}
+    URL_HASH SHA256=${CAT_SHA256}
+    DOWNLOAD_NO_PROGRESS TRUE
+    DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}
+    DOWNLOAD_NO_EXTRACT TRUE
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND ""
+    INSTALL_COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_BIN_DIR}
+    COMMAND ${CMAKE_COMMAND} -E copy <DOWNLOADED_FILE> ${DEPS_BIN_DIR})
```
Comment (justinmk) [inline cmake.deps/CMakeLists.txt:295]: would `GetBinaryDep` avoid some boilerplate here?

**Comment by justinmk** [comment]
>  unsure if this is even needed. We should be able to download the entire repo as an archive from github directly.

Could, though might be slow unless we do shallow/partial clone. Also, direct URLs makes these cases a bit less "special" I think?

File: cmake.deps/CMakeLists.txt (hunk 2 above)
Comment (dundargoc) [reply cmake.deps/CMakeLists.txt:295]: A variation of it, but in hindsight I don't think it's needed. A single download should be sufficient. I will investigate properly when I can.

**Comment by dundargoc** [comment]
> Could, though might be slow unless we do shallow/partial clone.

No, I meant the "archive" feature github has, to download any commit/tag as an archive like we do with some of our other dependencies like unibilium.

> Also, direct URLs makes these cases a bit less "special" I think?

Can you elaborate, missed what you mean.

**Comment by dundargoc** [comment]
The unstable PPA builds requires the `deps` repo to be as is, but it's probably too big to download the entire repo. Going back to the original plan of downloading required executables one by one.

File: cmake.deps/CMakeLists.txt (hunk 2 above)
Comment (dundargoc) [reply cmake.deps/CMakeLists.txt:296]: Fixed

**Comment by dundargoc** [comment]
@justinmk what about this? If it looks alright I can send PR to the deps repo.

File: cmake.deps/CMakeLists.txt [hunk 3]
```diff
@@ -163,9 +163,19 @@ set(LUV_SHA256 16720d17b9e6d42a7008b902207ea2d61e3eeaef3cf358398dbbf3777867ec09)
...
-# cat.exe curl.exe curl-ca-bundle.crt diff.exe tee.exe xxd.exe
-set(WINTOOLS_URL https://github.com/neovim/deps/raw/c1e7dd8de9e1b18d11dcfa0a192cd029262e5303/opt/win32tools.zip)
-set(WINTOOLS_SHA256 3c4c490a3d392ceeb1347cb77cc821a31900b688a2189276d3a1131a3f21daf1)
+# Windows only: cat.exe curl.exe curl-ca-bundle.crt diff.exe tee.exe xxd.exe
+set(CAT_URL https://github.com/dundargoc/deps/raw/master/opt/cat.exe)
+set(CAT_SHA256 93b8d307bb15af3968920bdea3beb869a49d166f9164853c58a4e6ffdcae61c6)
+set(CURL_CA_BUNDLE_URL https://github.com/dundargoc/deps/raw/master/opt/curl-ca-bundle.crt)
+set(CURL_CA_BUNDLE_SHA256 6ed95025fba2aef0ce7b647607225745624497f876d74ef6ec22b26e73e9de77)
+set(CURL_URL https://github.com/dundargoc/deps/raw/master/opt/curl.exe)
+set(CURL_SHA256 4c7151dea34c12b351030d4a14e50336a5a4e4639726565829997eea0d744654)
```
Comment (justinmk) [inline cmake.deps/CMakeLists.txt:172]: I vaguely recall curl involving several files. I assume you checked that `curl` works on windows with only these 2 files?

File: cmake.deps/cmake/GetBinaryDeps.cmake [hunk 4]
```diff
@@ -10,17 +10,10 @@ function(GetBinaryDep)
...
-  if(NOT _gettool_TARGET OR NOT _gettool_INSTALL_COMMAND)
-    message(FATAL_ERROR "Must pass INSTALL_COMMAND and TARGET")
-  endif()
-
   string(TOUPPER "${_gettool_TARGET}_URL" URL_VARNAME)
   string(TOUPPER "${_gettool_TARGET}_SHA256" HASH_VARNAME)
   set(URL ${${URL_VARNAME}})
   set(HASH ${${HASH_VARNAME}})
-  if(NOT URL OR NOT HASH )
-    message(FATAL_ERROR "${URL_VARNAME} and ${HASH_VARNAME} must be set")
```
Comment (justinmk) [inline cmake.deps/cmake/GetBinaryDeps.cmake:22]: Why isn't this needed/wanted?

**Review by justinmk — APPROVED** [review]
nice

File: cmake.deps/cmake/GetBinaryDeps.cmake (hunk 4 above)
Comment (dundargoc) [reply cmake.deps/cmake/GetBinaryDeps.cmake:22]: ExternalProject will emit the same warning anyway, so feels a bit unnecessary.

**Comment by dundargoc** [comment]
TODO: don't merge yet, need to change URL to the actual neovim dep url. Meed ro change the deps repo first.

**Comment by dundargoc** [comment]
> Windows 10 comes with the curl tool bundled with the operating system since version 1804.

Drop curl from neovim?

**Comment by dundargoc** [comment]
> Windows 10 comes with the curl tool bundled with the operating system since version 1804.

@justinmk thoughts?

**Review by justinmk — APPROVED** [review]
🚀

**Comment by clason** [comment]
@dundargoc are these reasonably up-to-date version, or should we bump them?

**Comment by dundargoc** [comment]
xxd is "version" 2022-01-14
cat is v2.1, latest is v9.2
tee doesn't have a version, it just says in the version description that "this version of tee is supplied with Vim 4.2 to make ':make' possible"
diff is v2.7 (Oct-1994), latest is 3.9

**Pull request merged**